components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt

```diff
@@ -18,77 +23,59 @@ import mozilla.components.feature.toolbar.internal.URLRenderer
 @Suppress("TooManyFunctions")
 class ToolbarPresenter(
     private val toolbar: Toolbar,
-    private val sessionManager: SessionManager,
-    private val sessionId: String? = null,
+    private val store: BrowserStore,
+    private val customTabId: String? = null,
     urlRenderConfiguration: ToolbarFeature.UrlRenderConfiguration? = null
-) : SelectionAwareSessionObserver(sessionManager) {
+) {
 
     @VisibleForTesting
     internal var renderer = URLRenderer(toolbar, urlRenderConfiguration)
 
+    private var subscription: Store.Subscription<BrowserState, BrowserAction>? = null
+
     /**
      * Start presenter: Display data in toolbar.
      */
     fun start() {
-        observeIdOrSelected(sessionId)
-        initializeView()
-
         renderer.start()
-    }
 
-    override fun stop() {
-        super.stop()
-
-        renderer.stop()
+        subscription = store.observeManually(onlyIfChanged<BrowserState, SessionState>(
+            onMainThread = true,
+            map = { state -> state.findCustomTabOrSelectedTab(customTabId) },
+            then = { _, tab -> render(tab) },
+            otherwise = { clear() }
+        ))
     }
 
     /**
-     * A new session has been selected: Update toolbar to display data of new session.
+     * Stop presenter: The toolbar will not be updated anymore.
      */
-    override fun onSessionSelected(session: Session) {
-        super.onSessionSelected(session)
-        initializeView()
-    }
-
-    override fun onAllSessionsRemoved() {
-        initializeView()
-    }
+    fun stop() {
+        subscription?.unsubscribe()
 
-    override fun onSessionRemoved(session: Session) {
-        if (sessionManager.selectedSession == null) {
-            initializeView()
-        }
+        renderer.stop()
     }
 
-    internal fun initializeView() {
-        val session = sessionId?.let { sessionManager.findSessionById(sessionId) }
-            ?: sessionManager.selectedSession
+    @VisibleForTesting(otherwise = PRIVATE)
+    internal fun render(tab: SessionState) {
+        renderer.post(tab.content.url)
 
-        renderer.post(session?.url ?: "")
+        toolbar.setSearchTerms(tab.content.searchTerms)
+        toolbar.displayProgress(tab.content.progress)
 
-        toolbar.setSearchTerms(session?.searchTerms ?: "")
-        toolbar.displayProgress(session?.progress ?: 0)
-        updateToolbarSecurity(session?.securityInfo ?: Session.SecurityInfo())
+        when (tab.content.securityInfo.secure) {
```

Comment:
Should just be an if statement.
```suggestion
        toolbar.siteSecure = if (tab.content.securityInfo.secure) SECURE else INSECURE
```

Reply:
:+1: